sinks/gcm/gcm_driver.go

```diff
@@ -0,0 +1,349 @@
+	"time"
+
+	"github.com/GoogleCloudPlatform/gcloud-golang/compute/metadata"
+)
+
+type MetricType int
+
+const (
+	// A cumulative metric.
+	MetricCumulative MetricType = iota
+
+	// An instantaneous value metric.
+	MetricGauge
+)
+
+func (self MetricType) String() string {
+	switch self {
+	case MetricCumulative:
+		return "cumulative"
+	case MetricGauge:
+		return "gauge"
+	}
+	return ""
+}
+
+type MetricValueType int
+
+const (
+	// An int64 value.
+	ValueInt64 MetricValueType = iota
+	// A boolean value
+	ValueBool
+	// A double-precision floating point number.
+	ValueDouble
+)
+
+func (self MetricValueType) String() string {
+	switch self {
+	case ValueInt64:
+		return "int64"
+	case ValueBool:
+		return "bool"
+	case ValueDouble:
+		return "double"
+	}
+	return ""
+}
+
+type LabelDescriptor struct {
+	// Key to use for the label.
+	Key string `json:"key,omitempty"`
+
+	// Description of the label.
+	Description string `json:"description,omitempty"`
+}
+
+type MetricDescriptor struct {
+	// The unique name of the metric.
+	Name string
+
+	// Description of the metric.
+	Description string
+
+	// Descriptor of the labels used by this metric.
+	Labels []LabelDescriptor
+
+	// Type and value of metric data.
+	Type      MetricType
+	ValueType MetricValueType
+}
+
+type Metric struct {
+	// The name of the metric. Must match an existing descriptor.
+	Name string
+
+	// The labels and values for the metric. The keys must match those in the descriptor.
+	Labels map[string]string
+
+	// The start and end time for which this data is representative.
+	Start time.Time
+	End   time.Time
+
+	// The value of the metric. Must match the type in the descriptor.
+	Value interface{}
+}
+
+type gcmDriver struct {
+	// Token to use for authentication.
+	token string
+
+	// When the token expires.
+	tokenExpiration time.Time
+
+	// TODO(vmarmol): Make this configurable and not only detected.
+	// GCE project.
+	project string
+
+	// TODO(vmarmol): Also store labels?
+	// Map of metrics we currently export.
+	exportedMetrics map[string]MetricDescriptor
+}
+
+// Returns a thread-compatible implementation of GCM interactions.
+func NewDriver() (*gcmDriver, error) {
+	// Only support GCE for now.
+	if !metadata.OnGCE() {
+		return nil, fmt.Errorf("the GCM sink is currently only supported on GCE")
+	}
+
+	// Detect project.
+	project, err := metadata.ProjectID()
+	if err != nil {
+		return nil, err
+	}
+
+	// Check required service accounts
+	err = checkServiceAccounts()
+	if err != nil {
+		return nil, err
+	}
+
+	impl := &gcmDriver{
+		project:         project,
+		exportedMetrics: make(map[string]MetricDescriptor),
+	}
+
+	// Get an initial token.
+	err = impl.refreshToken()
+	if err != nil {
+		return nil, err
+	}
+
+	return impl, nil
+}
+
+func (self *gcmDriver) refreshToken() error {
+	if time.Now().After(self.tokenExpiration) {
+		token, err := getToken()
+		if err != nil {
+			return nil
+		}
+
+		// Expire the token a bit early.
+		if token.ExpiresIn > 2 {
```

Comment:
nit: Add units here. As discussed offline, we should hopefully not have housekeeping intervals grater than `2 minutes`.